### PR TITLE
gitindex: replace go-git blob reading with pipelined git cat-file --batch

### PR DESCRIPTION
**Pull request by clemlesne**
## Summary

- Replace serial `go-git` `BlobObject` calls in `indexGitRepo` with a single pipelined `git cat-file --batch --buffer` subprocess
- A writer goroutine feeds all blob SHAs to stdin while the main goroutine reads responses from stdout, forming a concurrent pipeline
- `--buffer` switches git from per-object `write_or_die` flush to libc stdio buffering, reducing syscalls
- Submodule blobs fall back to the existing go-git `createDocument` path

## Benchmarks

On kubernetes (29,188 files, 261 MB content), Apple M1 Max, median of 5 runs:

| | go-git (before) | cat-file pipelined (after) | Change |
|---|---|---|---|
| **Time** | 3.05s | 0.62s | **-80% (4.9x)** |
| **Allocs** | 685K | 58K | **-92%** |
| **Memory** | 670 MB | 277 MB | **-59%** |

Reproducing:

```sh
git clone --depth=1 https://github.com/kubernetes/kubernetes /tmp/k8s
ZOEKT_BENCH_REPO=/tmp/k8s go test ./gitindex/ -bench='BenchmarkBlobRead_(GoGit|CatfilePipelined)/files=29188' -benchmem -count=5 -timeout=600s
```

## Context

Filed as [#1016](https://github.com/sourcegraph/zoekt/issues/1016). [seek](https://github.com/dualeai/seek) wraps zoekt as a local CLI for AI coding agents — every search is a blocking tool call, so cold-index speed matters.

Prior art: PR [#424](https://github.com/sourcegraph/zoekt/issues/424) benchmarked `git cat-file --batch` at 2.68x faster than go-git. PR [#940](https://github.com/sourcegraph/zoekt/pull/940) states "eventually we want to move away from go-git towards a vanilla direct git usage."

## Design

The pipelined reader (`readBlobsPipelined` in `gitindex/catfile.go`):

1. Starts `git cat-file --batch --buffer` with the repo as working directory
2. Writer goroutine feeds all blob SHAs via a 64KB `bufio.Writer`, then closes stdin to trigger git's final `fflush(stdout)`
3. Main goroutine reads responses in order via a 512KB `bufio.Reader`, parsing the `<oid> <type> <size>\n<content>\n` protocol
4. On early exit (read error), drains stdout so git can flush and exit, then waits for the writer goroutine and `cmd.Wait()`
5. Missing objects are marked via `blobResult.Missing` (matches existing `ErrObjectNotFound` → `skippedLargeDoc` behavior)

Integration in `indexGitRepo` (`gitindex/index.go`):

1. Separates main-repo keys (`SubRepoPath == ""`) from submodule keys
2. Collects blob SHAs in document-sorted order, calls `readBlobsPipelined` once
3. Builds `index.Document` from pipelined results for main-repo blobs
4. Falls back to go-git `createDocument` for submodule blobs

Per-blob allocations minimized to 1 (the content `[]byte` itself) via stack-allocated hex buffers, `ReadBytes` for headers, and `LastIndexByte` for size parsing.

## Test plan

- [x] `go test ./gitindex/ -race` passes (all existing tests + 2 new)
- [x] `go vet ./gitindex/` clean
- [x] `golangci-lint run ./gitindex/` — no new issues
- [x] New unit tests: `TestReadBlobsPipelined` (normal, empty, binary, large blobs), `TestReadBlobsPipelined_WithMissing`
- [x] New benchmarks: `BenchmarkBlobRead_GoGit` (baseline), `BenchmarkBlobRead_CatfilePipelined` (production path)
- [ ] Benchmark on a second large repo (e.g. `torvalds/linux`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: gitindex/index.go [hunk 1]
```diff
@@ -585,23 +585,78 @@ func indexGitRepo(opts Options, config gitIndexConfig) (bool, error) {
...
-	for idx, name := range names {
-		keys := fileKeys[name]
+	// Build the ordered list of (name, key) pairs and collect blob SHAs for
+	// the main repo so we can read them all at once via git cat-file --batch.
+	type indexEntry struct {
+		key       fileKey
+		blobIndex int // index into blobResults, or -1 for submodule blobs
+	}
+	entries := make([]indexEntry, 0, totalFiles)
+	mainRepoIDs := make([]plumbing.Hash, 0, totalFiles)
```
Comment (keegancsmith) [inline gitindex/index.go]: I think it makes sense to rather do two loops. One over mainRepoIDs and instead have another list for submoduleIDs. The actual code is quite minimal for submoduleIDs and the duplication then between what we do once we have created a doc is minimal.

File: gitindex/index.go [hunk 2]
```diff
@@ -585,23 +585,78 @@ func indexGitRepo(opts Options, config gitIndexConfig) (bool, error) {
...
+				entries = append(entries, indexEntry{key: key, blobIndex: -1})
+			}
+		}
+	}
+
+	// Bulk-read all main-repo blobs via pipelined cat-file --batch --buffer.
+	var blobResults []blobResult
+	if len(mainRepoIDs) > 0 {
+		var err error
+		blobResults, err = readBlobsPipelined(opts.RepoDir, mainRepoIDs)
```
Comment (keegancsmith) [inline gitindex/index.go]: I think there is a problem with this current API in that you read _everything_ into memory. This just won't scale for very large repos. It's part of why zoekt works with this design of adding documents that at certain thresholds decide to create a shard (which then rely on the GC to free memory).

Additionally previously if a blob was over our max size, we just wouldn't read it in to memory.

I would suggest instead designing the API to be like the `archive/tar` API. So you would still pass in a huge slice of mainRepoIDs. But it would then return a stateful struct over the output from the process. For each file we get a little bit of metadata (like size, etc) and then the caller here can read that into a []byte or call .Next to get it to skip over the file. 

So I think to keep the nice property around just passing in all IDs, we still get git to output all objects. But here in go land we decide between actually reading the object into a []byte or just io.Discarding it.

You don't have to go crazy on the design of the API. Just do something nice and minimal to suit this use case.

**Review by keegancsmith — CHANGES_REQUESTED** [review]
Awesome stuff. See inline, but the reason I am requesting changes is I want to avoid reading into memory files that we decide to not index due to being too large. I believe you should be able to shape the code written here into something which allows that.

File: gitindex/index.go (hunk 1 above)
Comment (clemlesne) [reply gitindex/index.go]: Done — split into two separate loops. `mainRepoKeys` iterated with the streaming reader, `submoduleKeys` iterated with go-git `createDocument`. The builder sorts documents internally via `sortDocuments` so ordering between the two loops does not matter.

File: gitindex/index.go (hunk 2 above)
Comment (clemlesne) [reply gitindex/index.go]: Agreed, reworked. `readBlobsPipelined` is replaced by a streaming `catfileReader` — `archive/tar.Reader`-style:

```go
cr, _ := newCatfileReader(opts.RepoDir, mainRepoIDs)
for _, key := range mainRepoKeys {
    size, missing, _ := cr.Next()
    if size > SizeMax { continue } // unread bytes auto-discarded by next Next()
    content := make([]byte, size)
    io.ReadFull(cr, content)
}
```

- `Next()` reads the header and returns size — content is not read until the caller calls `Read`
- Calling `Next()` again without reading discards unread bytes via `bufio.Reader.Discard`
- Large blobs never touch Go memory — only the header is parsed, content is discarded in the bufio buffer
- Peak memory is now bounded by `ShardMax` (one shard's worth of content), not total repo size

**Comment by clemlesne** [comment]
Updated benchmarks after the streaming rewrite (median of 5 runs, Apple M1 Max, kubernetes 29,188 files):

| | go-git (before) | catfileReader (after) | Change |
|---|---|---|---|
| **Time** | 2.41s | 0.52s | **-78% (4.6x)** |
| **Allocs** | 690K | 58K | **-92%** |
| **Memory** | 751 MB | 275 MB | **-63%** |

The streaming API is faster than the previous bulk `[]blobResult` approach (0.62s → 0.52s) because there's no intermediate slice — content goes directly from the pipe into the caller's buffer. And large blobs over `SizeMax` are now skipped without reading content into memory at all.

File: gitindex/catfile.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	if err := cmd.Start(); err != nil {
+		stdin.Close()
+		stdout.Close()
+		return nil, fmt.Errorf("start git cat-file: %w", err)
+	}
+
+	// Writer goroutine: feed all SHAs then close stdin to trigger flush.
+	writeErr := make(chan error, 1)
+	go func() {
+		defer stdin.Close()
```
Comment (keegancsmith) [inline gitindex/catfile.go:95]: also `defer close(writeErr)` so if you call Close twice it won't block forever.

File: gitindex/catfile.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+			return n, fmt.Errorf("read trailing LF: %w", err)
+		}
+		cr.pending = 0
+	}
+
+	return n, nil
+}
+
+// Close shuts down the cat-file process and waits for it to exit.
+func (cr *catfileReader) Close() error {
```
Comment (keegancsmith) [inline gitindex/catfile.go:201]: I worry about what happens when we call Close before git is finished giving us all its output. eg if we get an error during builder.Add we will call close before reading everything.

Maybe instead this should just always call something like cr.cmd.Kill() (I can't remember offhand what it is). Then I _think_ all the intermediatte goroutines/etc will drain.

I do quite like that you nicely clean up here, but I want to make sure we have good behaviour for early termination.

File: gitindex/catfile.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+	if bytes.HasSuffix(header, []byte(" missing")) {
+		return 0, true, nil
+	}
+
+	// Parse size from "<oid> <type> <size>".
+	lastSpace := bytes.LastIndexByte(header, ' ')
+	if lastSpace == -1 {
+		return 0, false, fmt.Errorf("unexpected header: %q", header)
+	}
+	size, err = strconv.ParseInt(string(header[lastSpace+1:]), 10, 64)
```
Comment (keegancsmith) [inline gitindex/catfile.go]: minor: lets parse this as an `int` instead of `int64` since all call sites work with int and then you can avoid the cast to int when you do Discard

File: gitindex/index.go [hunk 6]
```diff
@@ -585,25 +585,96 @@ func indexGitRepo(opts Options, config gitIndexConfig) (bool, error) {
...
-				return false, err
+				cr.Close()
+				return false, fmt.Errorf("cat-file next for %s: %w", key.FullPath(), err)
+			}
+
+			branches := repos[key].Branches
+			var doc index.Document
+
+			if missing {
+				doc = skippedLargeDoc(key, branches)
```
Comment (keegancsmith) [inline gitindex/index.go]: have you come across missing in your tests? Saying we skipped it due to large file doesn't seem correct. We should either fail or if this has a common reason add a new indexing reason for missing a document.

File: gitindex/index.go [hunk 7]
```diff
@@ -585,25 +585,96 @@ func indexGitRepo(opts Options, config gitIndexConfig) (bool, error) {
...
+			} else {
+				keyFullPath := key.FullPath()
+				if size > int64(opts.BuildOptions.SizeMax) && !opts.BuildOptions.IgnoreSizeMax(keyFullPath) {
+					// Skip without reading content into memory.
+					doc = skippedLargeDoc(key, branches)
+				} else {
+					content := make([]byte, size)
+					if _, err := io.ReadFull(cr, content); err != nil {
```
Comment (keegancsmith) [inline gitindex/index.go]: minor: unusual to repeatadly call io.ReadFull on the same reader. You might be able to expose a nicer api with the use of io.LimitedReader. But this is a very minor point, this does work!

File: gitindex/index.go [hunk 8]
```diff
@@ -585,25 +585,96 @@ func indexGitRepo(opts Options, config gitIndexConfig) (bool, error) {
...
+				submoduleKeys = append(submoduleKeys, key)
+			}
+		}
+	}
+
+	log.Printf("attempting to index %d total files (%d via cat-file, %d via go-git)", totalFiles, len(mainRepoIDs), len(submoduleKeys))
+
+	// Stream main-repo blobs via pipelined cat-file --batch --buffer.
+	// Large blobs are skipped without reading content into memory.
+	if len(mainRepoIDs) > 0 {
```
Comment (keegancsmith) [inline gitindex/index.go:618]: Feels a bit error prone the number of places we need to remember to do cr.Close inside of this if statement. Maybe this whole if statement body can be factored out into a function, then we can do defer cr.Close?

File: gitindex/index.go [hunk 9]
```diff
@@ -585,25 +585,96 @@ func indexGitRepo(opts Options, config gitIndexConfig) (bool, error) {
...
-		keys := fileKeys[name]
+	// Separate main-repo keys from submodule keys, collecting blob SHAs
+	// for the main repo so we can stream them via git cat-file --batch.
+	mainRepoKeys := make([]fileKey, 0, totalFiles)
+	mainRepoIDs := make([]plumbing.Hash, 0, totalFiles)
+	var submoduleKeys []fileKey
+
+	for _, name := range names {
+		for _, key := range fileKeys[name] {
+			if key.SubRepoPath == "" {
```
Comment (keegancsmith) [inline gitindex/index.go]: I have a slight concern this breaks stuff. For a few weeks can we have an envvar switch off this behaviour (I guess if turned off then if this if statement is always false it will just work out?)

Then after a few weeks we remove the envvar. But this code path is pretty critical on the server side both at Sourcegraph and GitLab.

**Review by keegancsmith — CHANGES_REQUESTED** [review]
Ok did a much deeper review now. It's looking great! Just have a few inline comments then we should be able to ship this.

File: gitindex/catfile.go (hunk 3 above)
Comment (clemlesne) [reply gitindex/catfile.go:95]: Done — added `defer close(writeErr)` in the goroutine. Also wrapped `Close()` itself in `sync.Once` (same pattern as `DirectoryWatcher.Stop` in `search/watcher.go`) so that `cmd.Wait()` is also called exactly once. This makes Close fully idempotent — safe to call multiple times or concurrently. Added `TestCatfileReader_DoubleClose` and `TestCatfileReader_ConcurrentClose` that verify this.

File: gitindex/catfile.go (hunk 4 above)
Comment (clemlesne) [reply gitindex/catfile.go:201]: Good call — switched to kill-first. `Close()` now calls `cmd.Process.Kill()` before draining. This matches Gitaly's pattern — their cache eviction code has an explicit comment: *"We first cancel the context such that the process gets a SIGKILL signal. Calling `close()` first may lead to a deadlock given that it waits for the process to exit, which may not happen if it hangs writing data to stdout."*

After Kill, the drain completes instantly (stdout pipe gets EOF), the writer goroutine gets EPIPE on its next write and exits, and `cmd.Wait()` returns. The expected "signal: killed" exit error is suppressed via `isKilledErr()`.

Added `TestCatfileReader_CloseBeforeExhausted_ManyBlobs` (200 files, reads only 1, then Close) and `TestCatfileReader_CloseWithoutReading` to exercise this.

File: gitindex/catfile.go (hunk 5 above)
Comment (clemlesne) [reply gitindex/catfile.go]: Done — `Next()` now returns `int`, `pending` field is `int`, uses `strconv.Atoi`. Removed all casts at call sites (`Discard`, size comparison against `SizeMax`, `make([]byte, size)`).

File: gitindex/index.go (hunk 6 above)
Comment (clemlesne) [reply gitindex/index.go]: You're right — added `SkipReasonMissing` with explanation `"object missing from repository"`. Missing blobs now get `skippedDoc(key, branches, index.SkipReasonMissing)` plus a warning log. This is unexpected for local repos (could indicate corruption, shallow clone, or a race with `git gc`), so the warning helps operators diagnose it.

Updated `file_category.go` to handle `SkipReasonMissing` the same way as `TooLarge` — guess category from filename, don't examine content.

The go-git `createDocument` path keeps using `SkipReasonTooLarge` for `ErrObjectNotFound` since that has a different semantic (large objects excluded during fetch).

File: gitindex/index.go (hunk 7 above)
Comment (clemlesne) [reply gitindex/index.go]: Considered this — `io.ReadFull` into a pre-allocated buffer is actually more efficient here since the size is known: zero intermediate allocations vs `io.ReadAll(io.LimitReader(...))` which grows a buffer internally. Added a comment explaining the choice. The `catfileReader` already limits reads internally via the `pending` field (same approach as `archive/tar.Reader`), so an extra `LimitedReader` wrapper would be redundant.

File: gitindex/index.go (hunk 8 above)
Comment (clemlesne) [reply gitindex/index.go:618]: Done — extracted `indexCatfileBlobs()` with `defer cr.Close()`. This eliminated all four manual `cr.Close()` calls (three error paths + one happy path). Any new error path added in the future is automatically covered by the defer.

File: gitindex/index.go (hunk 9 above)
Comment (clemlesne) [reply gitindex/index.go]: Done — added `ZOEKT_DISABLE_CATFILE_BATCH`. Follows the exact same pattern as `ZOEKT_DISABLE_GOGIT_OPTIMIZATION` already in this function: `cmp.Or(os.Getenv(...), "false")` + `strconv.ParseBool`. When set to `true`, the `if key.SubRepoPath == ""` condition is effectively always false, so all keys go through the go-git `createDocument` path. Benchmarked both paths on kubernetes — cat-file path is 4.4x faster, go-git fallback works correctly.

**Comment by clemlesne** [comment]
Addressed all inline comments from the second review — replied to each one individually. Summary of changes in `e0d964b`:

**Hardening:**
- `Close()` is now idempotent (`sync.Once`) and kills the git process first instead of draining, matching Gitaly's pattern. `defer close(writeErr)` prevents channel deadlock.
- Extracted `indexCatfileBlobs()` with `defer cr.Close()` — eliminates all manual Close calls.
- Added `ZOEKT_DISABLE_CATFILE_BATCH` env var kill switch (same pattern as `ZOEKT_DISABLE_GOGIT_OPTIMIZATION`).

**Correctness:**
- `Next()` returns `int` instead of `int64`, uses `strconv.Atoi`.
- New `SkipReasonMissing` for blobs git reports as missing (instead of reusing `SkipReasonTooLarge`).

**Testing:**
- 19 new hardening tests covering Close lifecycle, Read edge cases, missing objects, large blob byte precision, and more.
- All tests pass with `-race`, both with and without `ZOEKT_DISABLE_CATFILE_BATCH`.

**Performance:**
- Benchmarked on kubernetes (29,188 files) before and after — no regression (benchstat geomean -0.89%, within noise).
- With kill switch enabled, go-git fallback works correctly (4.4x slower as expected, but functionally equivalent).

File: gitindex/index.go (hunk 7 above)
Comment (keegancsmith) [reply gitindex/index.go]: What I meant is the catfileReader API exposes a LimitedReader rather than directly using it's Read implementation. But yeah, this all works so was more just me thinking it through. For example I think that is how the archive reader API likely works.

**Review by keegancsmith — APPROVED** [review]
